In [2]:
import pandas as pd  #importing pandas module 

In [3]:
df = pd.read_csv("titanic_train.csv")  # reading the csv file

In [4]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


## Column Descriptions

- **PassengerId** = Unique ID assigned to a passenger
- **Survived** = 0 for No, 1 for Yes
- **Pclass** = Passenger class: 1st, 2nd, or 3rd
- **Name** = Name of the passenger
- **Sex** = Male/Female
- **Age** = Age of the passenger
- **SibSp** = Number of siblings/spouses aboard
- **Parch** = Number of parents/children aboard
- **Ticket** = Ticket number
- **Fare** = Ticket fare
- **Cabin** = Cabin number
- **Embarked** = Port where the passenger boarded

In [5]:
df.shape # returns number of rows and coloumns this has 891 rows and 12 coloumns 

(891, 12)

In [6]:
df.columns  #returns the name fo all the diff coloumns in the dataset

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='str')

In [7]:
df.dtypes  # returns series containing the data types of each coloumn this series is a python object 

PassengerId      int64
Survived         int64
Pclass           int64
Name               str
Sex                str
Age            float64
SibSp            int64
Parch            int64
Ticket             str
Fare           float64
Cabin              str
Embarked           str
dtype: object

**DATA TYPES**

int64 -- whole number 
float64 -- decimals
object -- Pandas is treating the values as general objects, comonly text string data      

In [8]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB


In [9]:
missing = df.isnull().sum()  #returns a pandas series containing sum of missing values for each col
#calculating the number of missing values in each coloumn storing it in missing 
#missing stores a mapping structure like Age = 177
type(missing)  #it is a pandas series


pandas.Series

In [10]:
missing["Age"] # return the missing values 

np.int64(177)

In [11]:
missing[missing > 0]  # returns the coloums who's missing value is greater than 0

Age         177
Cabin       687
Embarked      2
dtype: int64

**FILTERING ROWS USING PANDAS**  
The Gneral formula is df[CONDITION] -- give me the rows from df where condition true
df["Age"] -- gives the coloumn
df["Age"].isnull() -- Checks every value in the Age Coloumn
                   -- Return true is value there and false if missing


In [12]:
df['Age']  #3 GIVES ALL THE ROWS BUT ONLY THE AGE COLOUMN

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888     NaN
889    26.0
890    32.0
Name: Age, Length: 891, dtype: float64

In [13]:
df["Age"].isnull() 

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888     True
889    False
890    False
Name: Age, Length: 891, dtype: bool

In [14]:
df[df["Age"].isnull()]  ## reading it "int to out " df["Age"] -- find the age coloumn
#df.["Age"].isnull() -- rows where age is null
#df[df.["Age"].isnull()] -- returns the rows which satisfy that condition 

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
17,18,1,2,"Williams, Mr. Charles Eugene",male,NaN,0,0,244373,13.0000,NaN,S
19,20,1,3,"Masselmani, Mrs. Fatima",female,NaN,0,0,2649,7.2250,NaN,C
26,27,0,3,"Emir, Mr. Farred Chehab",male,NaN,0,0,2631,7.2250,NaN,C
28,29,1,3,"O'Dwyer, Miss. Ellen ""Nellie""",female,NaN,0,0,330959,7.8792,NaN,Q
...,...,...,...,...,...,...,...,...,...,...,...,...
859,860,0,3,"Razi, Mr. Raihed",male,NaN,0,0,2629,7.2292,NaN,C
863,864,0,3,"Sage, Miss. Dorothy Edith ""Dolly""",female,NaN,8,2,CA. 2343,69.5500,NaN,S
868,869,0,3,"van Melkebeke, Mr. Philemon",male,NaN,0,0,345777,9.5000,NaN,S
878,879,0,3,"Laleff, Mr. Kristo",male,NaN,0,0,349217,7.8958,NaN,S


In [15]:
df[df["Age"].isnull()].describe() # SUMMARY OF ALL THE NUMERICAL COLOUMNS FOR 177 PASSENGERS 


,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,177.000000,177.000000,177.000000,0.0,177.000000,177.000000,177.000000
mean,435.581921,0.293785,2.598870,NaN,0.564972,0.180791,22.158567
std,250.552901,0.456787,0.763216,NaN,1.626316,0.534145,31.874608
min,6.000000,0.000000,1.000000,NaN,0.000000,0.000000,0.000000
25%,230.000000,0.000000,3.000000,NaN,0.000000,0.000000,7.750000
50%,452.000000,0.000000,3.000000,NaN,0.000000,0.000000,8.050000
75%,634.000000,1.000000,3.000000,NaN,0.000000,0.000000,24.150000
max,889.000000,1.000000,3.000000,NaN,8.000000,2.000000,227.525000


# count gives -- how many non missing values are there in that coloumn
# mean is average 
# 




In [16]:
df["Pclass"].value_counts() # returns a pandas series mapping each unique value to the number of time it appears 


Pclass
3    491
1    216
2    184
Name: count, dtype: int64

In [17]:
df[df["Age"].isnull()]['Pclass'].value_counts()
#1. Find rows where Age is missing
#2. Take only their Pclass
#3. Count how many are in each Pclass

Pclass
3    136
1     30
2     11
Name: count, dtype: int64

We know:

Entire Titanic dataset

There are 891 passengers:

Pclass 3 → 491
Pclass 1 → 216
Pclass 2 → 184
Passengers with missing Age

There are 177 passengers:

Pclass 3 → 136
Pclass 1 → 30
Pclass 2 → 11

Now, what are we comparing?

We're comparing the proportion of missing Age values in each class.

Why?

Because simply saying:

"136 missing-Age passengers are in Pclass 3"

doesn't tell us much.

There are already 491 total Pclass 3 passengers.

Maybe Pclass 3 just has more passengers overall, so naturally it has more missing ages.

In [18]:
df.groupby("Pclass")["Age"].apply(lambda x: x.isnull().mean())

Pclass
1    0.138889
2    0.059783
3    0.276986
Name: Age, dtype: float64

df.groupby("Pclass")["Age"].apply(lambda x: x.isnull().mean())

We're trying to answer:

For each Pclass, what percentage of Age values are missing?

Let's go piece by piece.

1. df.groupby("Pclass")

This means:

Split the DataFrame into groups based on Pclass.

So conceptually:

Pclass = 1
┌─────────────────────┐
│ passenger 1         │
│ passenger 2         │
│ passenger 3         │
└─────────────────────┘

Pclass = 2
┌─────────────────────┐
│ passenger 1         │
│ passenger 2         │
│ passenger 3         │
└─────────────────────┘

Pclass = 3
┌─────────────────────┐
│ passenger 1         │
│ passenger 2         │
│ passenger 3         │
└─────────────────────┘

We're basically saying:

"Separate my passengers into three groups: Pclass 1, Pclass 2, and Pclass 3."

2. ["Age"]

Now:

df.groupby("Pclass")["Age"]

means:

"After splitting passengers into Pclass groups, I only care about their Age column."

So instead of looking at every column:

Pclass 1 → all columns
Pclass 2 → all columns
Pclass 3 → all columns

we're focusing on:

Pclass 1 → Ages
Pclass 2 → Ages
Pclass 3 → Ages
3. .isnull()

Then:

df.groupby("Pclass")["Age"].apply(...)

Inside our function we're doing:

x.isnull()

Remember what .isnull() does?

It gives:

Age exists   → False
Age missing  → True

So for one group, it might look like:

Pclass 1:

False
False
True
False
True
False
...
4. .mean()

THIS is the clever part.

In Python:

False = 0
True  = 1

So if we have:

False
False
True
False
True

that's:

0
0
1
0
1

The mean is:

(0 + 0 + 1 + 0 + 1) / 5
= 2 / 5
= 0.40

So:

40% of the values are missing.

That's why we use:

x.isnull().mean()

We're calculating the proportion of missing values.

5. What's lambda x?

This:

lambda x: x.isnull().mean()

basically means:

"For each group, take its Age values (x), check which are missing, and calculate the average."

Don't worry about mastering lambda yet. We'll come back to it.

So the entire thing:
df.groupby("Pclass")["Age"].apply(lambda x: x.isnull().mean())

means:

Take DataFrame
      ↓
Group passengers by Pclass
      ↓
For each Pclass group
      ↓
Take Age
      ↓
Check which Age values are missing
      ↓
True = 1, False = 0
      ↓
Calculate the mean
      ↓
Get proportion missing

Result:

Pclass
1    0.138889
2    0.059783
3    0.277963

Which means approximately:

Pclass 1 → 13.9% of Age values missing
Pclass 2 →  6.0% of Age values missing
Pclass 3 → 27.8% of Age values missing

In [19]:
df.groupby("Pclass")["Age"].median()  ## finding the median age for each passenger class

Pclass
1    37.0
2    29.0
3    24.0
Name: Age, dtype: float64

Here's what that tells us:

Pclass	Typical Age (median)
1	37
2	29
3	24
Why is this useful?

Imagine we have a passenger like:

Pclass = 1
Age = NaN

If we simply used the overall median Age for everyone, we'd give them roughly the same age as a passenger from 3rd class.

But our data shows:

1st class → median age 37
2nd class → median age 29
3rd class → median age 24

So Pclass contains information about Age.

That means a more informed strategy could eventually be:

Age missing + Pclass 1 → use age around 37
Age missing + Pclass 2 → use age around 29
Age missing + Pclass 3 → use age around 24

This is called group-wise imputation.

BUT — we're still not doing it. 🚨

We have only investigated one variable.

Before deciding our production-quality imputation strategy, we should investigate other variables that could help predict Age, such as:

Sex
Pclass
SibSp
Parch

Maybe:

Pclass + Sex

gives us an even better estimate than Pclass alone.

For example, perhaps:

1st class + female → median Age 35
1st class + male   → median Age 40

That could be more informative.

So look at what we've actually accomplished

We started with:

Age → 177 missing

Instead of immediately doing:

df["Age"].fillna(...)

we investigated.

We discovered:

1. Missing Age isn't equally distributed across Pclass
       ↓
2. Pclass is associated with different typical ages
       ↓
3. Therefore Pclass could potentially help us estimate missing Age

That's the mindset I want you to learn.

We're not just learning Pandas commands.

We're learning:

"Before I modify data, I investigate the data and use evidence to choose the cleaning strategy."

In [20]:
df.groupby('Sex')['Age'].median()

Sex
female    27.0
male      29.0
Name: Age, dtype: float64

## Investigating Missing Age Using `SibSp` and `Parch`

We have **177 missing values in the `Age` column**.

Before filling them, we want to investigate whether other columns contain useful information that could help us estimate the missing ages.

The general approach is:

```text
Missing Age
    ↓
Look at other variables
    ↓
Check whether they are related to Age
    ↓
Check whether the relationship is reliable
    ↓
Use useful variables for imputation

In [21]:
df.groupby("SibSp")["Age"].median()

SibSp
0    29.0
1    30.0
2    23.0
3     9.5
4     6.5
5    11.0
8     NaN
Name: Age, dtype: float64

In [22]:
df.groupby("SibSp")["Age"].count()

SibSp
0    471
1    183
2     25
3     12
4     18
5      5
8      0
Name: Age, dtype: int64

1. Investigating SibSp

SibSp = number of siblings/spouses aboard.

We want to know:

Does the number of siblings/spouses aboard tell us anything useful about a passenger's age?

Step 1 — Find the median Age for each SibSp group
df.groupby("SibSp")["Age"].median()

This means:

groupby("SibSp") → split passengers into groups based on SibSp
["Age"] → look only at the Age column
.median() → calculate the median Age for each group

We got:

SibSp
0    29.0
1    30.0
2    23.0
3     9.5
4     6.5
5    11.0
8     NaN
What we noticed

There appears to be a relationship between SibSp and Age.

For example:

SibSp = 0 → median Age = 29
SibSp = 1 → median Age = 30
SibSp = 2 → median Age = 23
SibSp = 3 → median Age = 9.5
SibSp = 4 → median Age = 6.5
SibSp = 5 → median Age = 11

At first, this looks useful.

However, we cannot trust these medians without checking how many observations each median is based on.

Step 2 — Check the number of known Ages in each SibSp group
df.groupby("SibSp")["Age"].count()

.count() counts the non-missing Age values in each group.

We got:

SibSp
0    471
1    183
2     25
3     12
4     18
5      5
8      0
Why did we check the count?

Because a median calculated from a large group is generally more reliable than one calculated from a tiny group.

For example:

SibSp = 0
471 known Ages
median = 29

This is based on a large amount of data.

But:

SibSp = 5
5 known Ages
median = 11

This is based on only 5 observations, so we should be cautious about using 11 as a representative age.

The most extreme example is:

SibSp = 8
0 known Ages
median = NaN

There are no known Ages in this group, so a median cannot be calculated.

Conclusion about SibSp

SibSp appears to contain some information about Age, but its groups are uneven in size.

Some groups have plenty of observations, while others have very few or none.

Therefore:

SibSp may be useful as an additional feature for estimating Age, but using SibSp alone to fill missing Age values could be unreliable.

In [23]:
df.groupby("Parch")["Age"].median()

Parch
0    30.0
1    23.0
2    16.5
3    24.0
4    42.5
5    39.0
6    43.0
Name: Age, dtype: float64

In [24]:
df.groupby("Parch")["Age"].count()

Parch
0    521
1    110
2     68
3      5
4      4
5      5
6      1
Name: Age, dtype: int64

2. Investigating Parch

Parch = number of parents/children aboard.

We use the same investigation strategy.

We want to know:

Does the number of parents/children aboard tell us anything useful about a passenger's age?

Step 1 — Find the median Age for each Parch group
df.groupby("Parch")["Age"].median()

This means:

Group passengers by Parch
Select the Age column
Calculate the median Age for each group

We got:

Parch
0    30.0
1    23.0
2    16.5
3    24.0
4    42.5
5    39.0
6    43.0

There appears to be some variation in Age between the different Parch groups.

However, again:

A median by itself is not enough.

We need to know how many known Age values were used to calculate each median.

Step 2 — Check the number of known Ages in each Parch group
df.groupby("Parch")["Age"].count()

This counts the non-missing Age values in each Parch group.

We got:

Parch
0    521
1    110
2     68
3      5
4      4
5      5
6      1
Why this matters

Look at:

Parch = 6
Median Age = 43
Known Ages = 1

There is only one passenger with a known Age in this group.

Therefore, using 43 as a reliable typical age for Parch = 6 would be a bad assumption.

Similarly:

Parch = 4
Median Age = 42.5
Known Ages = 4

and:

Parch = 3
Median Age = 24
Known Ages = 5

These groups are also very small.

Compare them with:

Parch = 0
Median Age = 30
Known Ages = 521

and:

Parch = 1
Median Age = 23
Known Ages = 110

These medians are based on much more data.

Important Lesson: Always Check Sample Size

A group statistic such as a median does not automatically mean it is reliable.

For example:

Group A:
median = 30
observations = 500

is generally much more informative than:

Group B:
median = 43
observations = 1

The second number may simply describe one unusual individual rather than a meaningful pattern.

Therefore, when investigating relationships between variables, we should look at:

Statistic
+
Number of observations

not just the statistic by itself.

Comparison: SibSp vs Parch
SibSp
SibSp = 0 → median Age 29.0 → 471 known Ages
SibSp = 1 → median Age 30.0 → 183 known Ages
SibSp = 2 → median Age 23.0 → 25 known Ages
SibSp = 3 → median Age 9.5  → 12 known Ages
SibSp = 4 → median Age 6.5  → 18 known Ages
SibSp = 5 → median Age 11.0 → 5 known Ages
SibSp = 8 → NaN             → 0 known Ages
Parch
Parch = 0 → median Age 30.0 → 521 known Ages
Parch = 1 → median Age 23.0 → 110 known Ages
Parch = 2 → median Age 16.5 → 68 known Ages
Parch = 3 → median Age 24.0 → 5 known Ages
Parch = 4 → median Age 42.5 → 4 known Ages
Parch = 5 → median Age 39.0 → 5 known Ages
Parch = 6 → median Age 43.0 → 1 known Age

In [25]:
df.groupby(["Pclass", "Sex"])["Age"].median()

Pclass  Sex   
1       female    35.0
        male      40.0
2       female    28.0
        male      30.0
3       female    21.5
        male      25.0
Name: Age, dtype: float64

In [26]:
df.groupby(["Pclass", "Sex"])["Age"].count()

Pclass  Sex   
1       female     85
        male      101
2       female     74
        male       99
3       female    102
        male      253
Name: Age, dtype: int64

In [27]:
df_clean=df.copy()

In [28]:
df_clean["Age"] = df_clean.groupby(["Pclass", "Sex"])["Age"].transform(
    lambda x: x.fillna(x.median())
)

Our code:

df_clean["Age"] = df_clean.groupby(["Pclass", "Sex"])["Age"].transform(
    lambda x: x.fillna(x.median())
)

Think of it as:

df_clean
   ↓
group by Pclass + Sex
   ↓
take Age
   ↓
for each group:
    find median Age
    fill missing Ages with that median
   ↓
return values aligned with original rows
   ↓
put them back into df_clean["Age"]
1. Start with this
df_clean

That's our cleaning copy of the original DataFrame.

2. Group by Pclass and Sex
df_clean.groupby(["Pclass", "Sex"])

We're saying:

"Split the passengers into groups based on BOTH Pclass and Sex."

So we get groups like:

Pclass 1 + female
Pclass 1 + male
Pclass 2 + female
Pclass 2 + male
Pclass 3 + female
Pclass 3 + male
3. Select Age
df_clean.groupby(["Pclass", "Sex"])["Age"]

Now we're saying:

"Okay, I created those groups. I only care about their Age values."

Conceptually:

Pclass 1 + female → Ages
Pclass 1 + male   → Ages
Pclass 2 + female → Ages
...
4. .transform(...)

Now comes the important part:

.transform(...)

transform() says:

"Perform this operation separately on each group, but give me the result back in the same shape/order as the original data."

That's extremely useful here because eventually we need one Age value for every passenger.

5. The lambda

Inside transform() we have:

lambda x: x.fillna(x.median())

You can temporarily read:

lambda x:

as:

"For each group, call that group's Age values x and do this..."

So imagine we're processing:

Pclass = 3
Sex = female

and its Ages are:

21
NaN
18
25
NaN
30

That entire group becomes x.

6. x.median()
x.median()

Calculate the median of that group's known ages.

Let's pretend the median is:

21.5
7. x.fillna(...)

Now:

x.fillna(x.median())

means:

Replace the missing values in this group with the group's median.

So:

Before:

21
NaN
18
25
NaN
30

After:

21
21.5
18
25
21.5
30

Existing ages are not changed.

Only NaN values are filled.

8. Why transform() instead of just median()?

This is the really important part.

If we did:

df_clean.groupby(["Pclass", "Sex"])["Age"].median()

we'd get:

Pclass  Sex
1       female    35
1       male      40
2       female    28
...

That's only 6 median values.

But our DataFrame has 891 passengers.

We need a result with 891 positions, one for each passenger.

That's what transform() gives us.

Conceptually:

Original DataFrame
891 rows
     ↓
groupby
     ↓
6 groups
     ↓
calculate group median
     ↓
transform back to 891 rows

So every passenger gets associated with the median of their own group.

9. Finally, the assignment

The whole right side:

df_clean.groupby(["Pclass", "Sex"])["Age"].transform(
    lambda x: x.fillna(x.median())
)

produces a Series containing the updated Age values for all 891 rows.

Then:

df_clean["Age"] =

puts those values back into the Age column.

So the whole thing is basically:

                df_clean
                   ↓
        group by Pclass + Sex
                   ↓
             get Age values
                   ↓
          calculate each group's
             median Age
                   ↓
        fill only the NaN Ages
                   ↓
       produce 891 Age values
                   ↓
       replace df_clean["Age"]
The one sentence to remember

groupby() creates the groups, transform() applies a group-level operation while keeping the original row structure, fillna() replaces missing values, and median() provides the replacement value.

And you did the right thing by making df_clean = df.copy() first — your original raw df remains untouched.

In [29]:
df_clean["Cabin"].head(20)

0      NaN
1      C85
2      NaN
3     C123
4      NaN
5      NaN
6      E46
7      NaN
8      NaN
9      NaN
10      G6
11    C103
12     NaN
13     NaN
14     NaN
15     NaN
16     NaN
17     NaN
18     NaN
19     NaN
Name: Cabin, dtype: str

In [33]:
df_clean["Cabin"].dropna().head(20) #Take the cabin coloumn -- Drop the null values -- return the actual top 20 values 

1             C85
3            C123
6             E46
10             G6
11           C103
21            D56
23             A6
27    C23 C25 C27
31            B78
52            D33
54            B30
55            C52
61            B28
62            C83
66            F33
75          F G73
88    C23 C25 C27
92            E31
96             A5
97        D10 D12
Name: Cabin, dtype: str

There are actually two pieces of information hidden inside Cabin:

C85
││
│└── cabin number
└── deck letter

So:

C85  → Deck C, Cabin 85
E46  → Deck E, Cabin 46
G6   → Deck G, Cabin 6

And then we have passengers with multiple cabins:

C23 C25 C27

That means the Cabin column isn't just a simple categorical variable like:

male / female

It contains structured information.

In [ ]:
df_clean["Cabin"].str[0].value_counts()
#Take the first letter of every non-missing Cabin value and count how many passengers have each deck letter.

Cabin
C    59
B    47
D    33
E    32
A    15
F    13
G     4
T     1
Name: count, dtype: int64

In [ ]:
df_clean["Cabin"].isnull().mean() #Calculate the proportion of rows where Cabin is missing.
#is null returns TRUE/FALSE when mean is applied pandas treats true as 1 and false as 0

#77% values are Null

np.float64(0.7710437710437711)

In [ ]:
df_clean.groupby('Pclass')['Cabin'].apply(lambda x: x.isnull().mean())
#Cabin information is heavily concentrated among 1st-class passengers, while Cabin values are missing for the vast majority of 2nd- and 3rd-class passengers.

Pclass
1    0.185185
2    0.913043
3    0.975560
Name: Cabin, dtype: float64

## Investigating Missing Cabin Values by Pclass

We know that the `Cabin` column has a large number of missing values:

- Total rows: 891
- Missing Cabin values: 687
- Missing percentage: approximately 77.1%

Before deciding what to do with these missing values, we want to investigate whether the missingness is related to another variable.

We will investigate whether the percentage of missing `Cabin` values is different across the three passenger classes (`Pclass`).

### Command

df_clean.groupby("Pclass")["Cabin"].apply(lambda x: x.isnull().mean())

### What does this command do?

The entire DataFrame is first divided into groups based on `Pclass`.

Since `Pclass` has three possible values, we conceptually get three groups:

Entire DataFrame
↓
Group by Pclass
↓
Pclass = 1
Pclass = 2
Pclass = 3

We then select the `Cabin` column from each group.

Conceptually:

Pclass = 1 → Cabin values
Pclass = 2 → Cabin values
Pclass = 3 → Cabin values

The `["Cabin"]` part tells Pandas that we only want to work with the Cabin column.

### Understanding `apply()`

`.apply()` tells Pandas to run the given function separately on each Pclass group.

The function is:

lambda x: x.isnull().mean()

Here, `x` represents the `Cabin` Series belonging to the current Pclass group.

So conceptually:

Pclass = 1 → x = Cabin values for Pclass 1
Pclass = 2 → x = Cabin values for Pclass 2
Pclass = 3 → x = Cabin values for Pclass 3

### Understanding `isnull()`

The next operation is:

x.isnull()

This checks every Cabin value and returns a Boolean value.

If the Cabin value exists:

False

If the Cabin value is missing:

True

For example:

C85 → False
NaN → True
E46 → False
NaN → True
NaN → True

So `isnull()` converts the Cabin column into a Series of True and False values.

### Understanding `mean()`

We then use:

x.isnull().mean()

When calculating the mean of Boolean values, Pandas treats:

True = 1
False = 0

For example:

True
False
True
False
True

becomes:

1
0
1
0
1

The mean is therefore:

(1 + 0 + 1 + 0 + 1) / 5

= 3 / 5

= 0.60

Therefore:

0.60 = 60%

So the mean gives us the proportion of missing Cabin values.

### What are we calculating?

Because the operation is performed separately for each Pclass group, we get:

Pclass = 1 → percentage of missing Cabin values
Pclass = 2 → percentage of missing Cabin values
Pclass = 3 → percentage of missing Cabin values

This allows us to compare Cabin missingness across the three passenger classes.

### Overall Flow

Entire DataFrame
↓
Group by Pclass
↓
Three groups:
    Pclass = 1
    Pclass = 2
    Pclass = 3
↓
Select Cabin
↓
Apply function to each group
↓
isnull()
↓
True / False
↓
True = 1
False = 0
↓
mean()
↓
Proportion of missing Cabin values
↓
Compare missing Cabin percentage across Pclass

### One-Sentence Summary

We divide the dataset into three groups based on `Pclass`, select the `Cabin` column from each group, use `isnull()` to identify missing Cabin values as True/False, and calculate the mean of those Boolean values to determine the proportion of missing Cabin values within each passenger class.

### Why are we doing this?

We are trying to understand whether the missingness in `Cabin` is random or whether it is related to `Pclass`.

If, for example, first-class passengers have much less missing Cabin information than third-class passengers, then the missingness itself may contain useful information.

This is important because we should understand the missing data before deciding whether to:

- Fill the missing values
- Replace them with `"Unknown"`
- Extract another feature such as Deck
- Create a missing-value indicator
- Drop the original `Cabin` column
- Use another preprocessing strategy

We are currently in the investigation stage.

Detect → Investigate → Understand → Decide → Clean → Validate

In [40]:
df_clean.groupby(df_clean["Cabin"].isnull())["Survived"].mean()
#df_clean["Cabin"].isnull() -- itself produces a Series of True/False values with the same 891 rows
#So Pandas can use those True/False values as the grouping labels.
#We're essentially saying:
#Put every passenger whose Cabin is missing into one group, and every passenger whose Cabin exists into another group.
#then calculating the survival rate inside these 2 groups 


Cabin
False    0.666667
True     0.299854
Name: Survived, dtype: float64

Cabin exists     → 66.7% survived
Cabin missing    → 30.0% survived
passengers with a recorded Cabin had a much higher survival rate.

but we already found
Pclass 1 → 18.5% Cabin missing → 81.5% Cabin recorded
Pclass 2 → 91.3% Cabin missing →  8.7% Cabin recorded
Pclass 3 → 97.6% Cabin missing →  2.4% Cabin recorded

having a recorded Cabin is strongly associated with being 1st class.

And we know from the Titanic data that Pclass itself is strongly associated with survival.

In [ ]:
df_clean.groupby(["Pclass",df_clean['Cabin'].isnull()])["Survived"].mean()

Pclass  Cabin
1       False    0.664773
        True     0.475000
2       False    0.812500
        True     0.440476
3       False    0.500000
        True     0.235908
Name: Survived, dtype: float64

```markdown
## Investigating Whether Cabin Missingness Is Informative

We found that the `Cabin` column has a very large amount of missing data:

- Total rows: 891
- Missing Cabin values: 687
- Missing percentage: approximately 77.1%

Because so much of the `Cabin` column is missing, we need to understand whether the missingness itself contains useful information.

---

### 1. Cabin Missingness and Overall Survival

We first compared survival rates between passengers who have Cabin information and passengers whose Cabin information is missing.

Command:

df_clean.groupby(df_clean["Cabin"].isnull())["Survived"].mean()

Result:

False    0.666667
True     0.299854

Remember:

False → Cabin exists
True  → Cabin is missing

Therefore:

Cabin exists → approximately 66.7% survived
Cabin missing → approximately 30.0% survived

At first, this suggests that having Cabin information is associated with a much higher survival rate.

However, this alone does NOT prove that Cabin information itself is responsible for the higher survival rate.

---

### 2. Relationship Between Cabin Availability and Pclass

We previously found that Cabin availability is strongly related to passenger class.

Missing Cabin percentages by `Pclass`:

Pclass 1 → 18.5% missing
Pclass 2 → 91.3% missing
Pclass 3 → 97.6% missing

Therefore:

Pclass 1 → most passengers have Cabin information
Pclass 2 → most passengers do not have Cabin information
Pclass 3 → almost all passengers do not have Cabin information

This means that `Cabin` availability is strongly associated with `Pclass`.

Because `Pclass` is also related to survival, the higher survival rate among passengers with Cabin information could partly be explained by the fact that passengers with Cabin information are much more likely to be first class.

Therefore, we should not immediately conclude:

"Having a Cabin causes a passenger to have a higher chance of survival."

---

### 3. Control for Pclass

To investigate whether Cabin missingness still contains information after accounting for passenger class, we compared survival rates within each `Pclass`.

Command:

df_clean.groupby(["Pclass", df_clean["Cabin"].isnull()])["Survived"].mean()

Result:

Pclass  Cabin    Survival Rate

1       False    0.664773
1       True     0.475000

2       False    0.812500
2       True     0.440476

3       False    0.500000
3       True     0.235908

Remember:

False → Cabin exists
True  → Cabin is missing

---

### 4. Interpretation

Now we are comparing passengers within the same passenger class.

#### Pclass 1

Cabin exists → approximately 66.5% survived

Cabin missing → approximately 47.5% survived

#### Pclass 2

Cabin exists → approximately 81.3% survived

Cabin missing → approximately 44.0% survived

#### Pclass 3

Cabin exists → 50.0% survived

Cabin missing → approximately 23.6% survived

The difference remains within all three passenger classes.

This is important because it means the difference in survival cannot be explained by `Pclass` alone.

For example, among third-class passengers:

Cabin exists → 50.0% survived

Cabin missing → 23.6% survived

Both groups are Pclass 3, but their survival rates are still substantially different.

---

### 5. What We Have Learned

Our investigation suggests that Cabin missingness itself contains useful information.

The relationship can be visualized conceptually as:

Cabin availability
        ↓
Strongly related to Pclass
        ↓
Pclass is related to Survival

BUT:

Even within the same Pclass:

Cabin exists
        ↓
Higher survival rate

Cabin missing
        ↓
Lower survival rate

Therefore, Cabin missingness appears to contain information beyond Pclass alone.

---

### 6. Important Warning

We should NOT conclude:

"Having a Cabin causes higher survival."

We have only found an association.

There may be other variables responsible for part of this relationship.

For example, Cabin availability could be related to:

- Exact location on the ship
- Passenger type
- Family characteristics
- Socioeconomic factors
- Whether the passenger was recorded in the cabin information
- Other variables in the dataset

Therefore:

Correlation or association does not automatically mean causation.

---

### 7. Preprocessing Implication

Because the fact that Cabin is missing appears to contain useful information, we should NOT blindly replace every missing Cabin value with an arbitrary value and then forget that it was originally missing.

Instead, one possible strategy is to create a separate feature that records whether Cabin information exists.

For example:

df_clean["CabinKnown"] = df_clean["Cabin"].notna().astype(int)

This would produce:

Cabin exists  → 1
Cabin missing → 0

The new feature preserves the information contained in the missingness itself.

We can then decide what to do with the original `Cabin` column separately.

---

### Current Status

We have NOT yet modified the `Cabin` column.

So far, we have only investigated it.

Our workflow is:

Detect
↓
Investigate
↓
Understand
↓
Decide
↓
Clean
↓
Validate

We are currently at the:

**Understand → Decide**

stage.

---

### Key Lesson

Missing data is not always just a problem that needs to be removed.

Sometimes:

> **The fact that a value is missing can itself contain useful information.**

In this dataset, `Cabin` missingness is strongly related to `Pclass` and is also associated with survival even after grouping by `Pclass`.

Therefore, the missingness of `Cabin` should be treated as potentially informative rather than automatically discarded.
```


In [ ]:
##1. Create a CabinKnown feature

##This preserves whether Cabin information existed: Cabin exists  → 1 Cabin missing → 0
df_clean["cabinknown"] = df_clean["Cabin"].notna().astype(int) #notna()  → True when present

In [44]:
##CHECKING HOW MANY PASSENGERS HAVE MORE THAN ONE CABIN
df_clean.loc[df_clean["Cabin"].str.contains(" ",na=False), "Cabin"]

27         C23 C25 C27
75               F G73
88         C23 C25 C27
97             D10 D12
118            B58 B60
128              F E69
297            C22 C26
299            B58 B60
305            C22 C26
311    B57 B59 B63 B66
341        C23 C25 C27
390            B96 B98
435            B96 B98
438        C23 C25 C27
498            C22 C26
679        B51 B53 B55
699              F G63
700            C62 C64
715              F G73
742    B57 B59 B63 B66
763            B96 B98
789            B82 B84
802            B96 B98
872        B51 B53 B55
Name: Cabin, dtype: str

## Finding Passengers With Multiple Cabin Entries Using `.loc`

We found that some passengers have more than one cabin recorded in the `Cabin` column.

For example:

C23 C25 C27
D10 D12

We want to find all passengers whose `Cabin` value contains multiple cabin entries.

There are 24 such passengers in this dataset.

### Command

df_clean.loc[df_clean["Cabin"].str.contains(" ", na=False), "Cabin"]

---

## Breaking the Command Down

The command can be understood from the inside out:

df_clean.loc[condition, "Cabin"]

The `.loc` accessor allows us to select specific rows and columns from a DataFrame using labels or Boolean conditions.

The general structure is:

df.loc[rows, columns]

In our case:

df_clean.loc[condition, "Cabin"]

means:

> Select the rows where `condition` is True, and from those rows return only the `Cabin` column.

---

## 1. Start With `df_clean`

df_clean

This is our cleaned DataFrame containing the Titanic dataset.

---

## 2. Select the `Cabin` Column

df_clean["Cabin"]

This selects only the `Cabin` column from the DataFrame.

Conceptually, we might have:

NaN
C85
NaN
C123
C23 C25 C27
B78
D10 D12
...

This gives us a Pandas Series containing the Cabin values.

---

## 3. `.str.contains(" ")`

Next we have:

df_clean["Cabin"].str.contains(" ")

`.str` allows us to perform string operations on every value in a Pandas Series.

`.contains(" ")` checks whether each string contains the specified value.

The specified value here is:

" "

which is simply a space.

Why are we looking for a space?

Because multiple cabin entries are separated by spaces.

For example:

C23 C25 C27

contains spaces, while:

C85

does not.

Therefore:

C85 → False

C123 → False

C23 C25 C27 → True

D10 D12 → True

---

## 4. What Does `na=False` Mean?

Some Cabin values are missing:

NaN

If we run:

df_clean["Cabin"].str.contains(" ")

without `na=False`, missing values can produce missing/undefined results instead of a clean Boolean value.

By writing:

df_clean["Cabin"].str.contains(" ", na=False)

we are telling Pandas:

> If the Cabin value is missing, treat the result as False.

So we get a clean Boolean Series:

Cabin value       Result

C85               False
C123              False
NaN               False
C23 C25 C27       True
D10 D12           True

This is useful because we only want to identify actual Cabin strings containing spaces.

---

## 5. The Condition

This entire part:

df_clean["Cabin"].str.contains(" ", na=False)

creates a Boolean Series.

Conceptually:

True  → Cabin contains a space
False → Cabin does not contain a space

For example:

False
False
False
True
False
True
...

This Boolean Series has the same number of rows as the original DataFrame.

---

## 6. Using `.loc`

Now we use:

df_clean.loc[condition, "Cabin"]

The general structure of `.loc` is:

df.loc[rows, columns]

The first part specifies which rows we want.

The second part specifies which columns we want.

So:

df_clean.loc[condition, "Cabin"]

means:

> Give me the rows where the condition is True, and only show me the `Cabin` column.

---

## 7. Putting Everything Together

Our complete command is:

df_clean.loc[df_clean["Cabin"].str.contains(" ", na=False), "Cabin"]

We can read it as:

> From `df_clean`, find every row where the `Cabin` value contains a space, and return only the `Cabin` values from those rows.

Conceptually:

Entire DataFrame
        ↓
Select Cabin
        ↓
Check each Cabin for " "
        ↓
Create True / False condition
        ↓
Use the condition to filter rows
        ↓
Return only the Cabin column
        ↓
Get passengers with multiple cabin entries

---

## 8. Why Are We Doing This?

We are investigating the structure of the `Cabin` column before extracting the deck.

We already discovered values such as:

C85
E46
G6
B78

For these values, extracting the first character is straightforward:

C85 → C
E46 → E
G6  → G
B78 → B

However, some passengers have multiple cabins:

C23 C25 C27
D10 D12

We need to understand these cases before deciding how to create a `Deck` feature.

For example:

C23 C25 C27

contains three cabin numbers, but all three belong to deck C.

Therefore:

C23 C25 C27 → C

would be reasonable.

However, if a passenger had:

C23 E46

then simply extracting the first character would give:

C

and we would lose the information that the passenger also had a cabin on deck E.

Therefore, we first investigate the 24 passengers with multiple cabin entries.

---

## Important `.loc` Pattern

The pattern:

df.loc[condition, "Column"]

is extremely important in Pandas.

It means:

> Select rows based on a condition and select a specific column.

For example:

df_clean.loc[df_clean["Pclass"] == 1, "Age"]

means:

> Select the `Age` values of passengers whose `Pclass` is 1.

Similarly:

df_clean.loc[df_clean["Cabin"].str.contains(" ", na=False), "Cabin"]

means:

> Select the `Cabin` values of passengers whose Cabin entry contains a space.

---

## One-Sentence Summary

> `.loc[condition, "Cabin"]` filters the DataFrame to the rows where the specified Boolean condition is True and returns only the `Cabin` column; in our case, we use `.str.contains(" ", na=False)` to find passengers whose Cabin field contains multiple cabin entries.

In [45]:
## EXTRACTING THE DECK LETTERS
df_clean["Cabin"].str.findall(r"[A-G]")

0      NaN
1      [C]
2      NaN
3      [C]
4      NaN
      ... 
886    NaN
887    [B]
888    NaN
889    [C]
890    NaN
Name: Cabin, Length: 891, dtype: object

In [47]:
# ONE PROBLEM WE HAVE EXTRACTED THE DECKS BUT FOR MULTIPLE DECKS LIKE C85 C76 C87
#C IS EXTRACKED THRICE [C,C,C] we need to remove the deuplicated right

df_clean["Cabin"].str.findall(r"[A-G]").apply(lambda x : list(dict.fromkeys(x)) if isinstance(x,list) else x )

0      NaN
1      [C]
2      NaN
3      [C]
4      NaN
      ... 
886    NaN
887    [B]
888    NaN
889    [C]
890    NaN
Name: Cabin, Length: 891, dtype: object

## Handling Duplicate Deck Letters in Cabin Values

We previously used:

df_clean["Cabin"].str.findall(r"[A-G]")

This extracts every deck letter from each Cabin value.

For example:

C23 C25 C27 → ["C", "C", "C"]

F G73 → ["F", "G"]

F E69 → ["F", "E"]

A missing Cabin value → NaN

The problem is that when multiple cabins are on the same deck, we get duplicate letters.

For example:

C23 C25 C27 → ["C", "C", "C"]

We only need to record the deck once:

["C"]

However, we must preserve different decks.

For example:

F G73 → ["F", "G"]

F E69 → ["F", "E"]

Therefore, we need to remove duplicate deck letters while keeping different deck letters.

---

## 1. `dict.fromkeys(x)`

The syntax is:

dict.fromkeys(iterable)

It creates a dictionary using the values from `x` as dictionary keys.

Dictionary keys must be unique, so duplicate values are automatically removed.

For example:

x = ["C", "C", "C"]

dict.fromkeys(x)

produces:

{"C": None}

The duplicate `"C"` values disappear because dictionary keys must be unique.

Another example:

x = ["F", "G", "F"]

dict.fromkeys(x)

produces:

{"F": None, "G": None}

The second `"F"` is removed.

---

## 2. `list(...)`

We do not actually want a dictionary as our final result.

We want a list of deck letters.

So we use:

list(dict.fromkeys(x))

The process is:

["C", "C", "C"]
↓
dict.fromkeys(x)
↓
{"C": None}
↓
list(...)
↓
["C"]

Therefore:

dict.fromkeys(x) → removes duplicates

list(...) → converts the result back into a list

So:

["C", "C", "C"] → ["C"]

and:

["F", "G"] → ["F", "G"]

---

## 3. Understanding `isinstance()`

The syntax is:

isinstance(object, type)

It asks:

> Is this object an instance of this type?

For example:

x = ["C", "C", "C"]

isinstance(x, list)

returns:

True

because `x` is a list.

If:

x = "C85"

isinstance(x, list)

returns:

False

because `"C85"` is a string, not a list.

In our case:

isinstance(x, list)

means:

> Check whether `x` is a list.

---

## 4. Why Do We Need `isinstance()`?

Our `str.findall()` result contains two different types of values.

Passengers with Cabin information have lists:

["C"]

["C", "C", "C"]

["F", "G"]

Passengers with missing Cabin information have:

NaN

Therefore, we cannot blindly run:

list(dict.fromkeys(x))

on every value.

We only want to run it when `x` is actually a list.

So we use:

isinstance(x, list)

to check the type of `x`.

---

## 5. The `if / else` Expression

The syntax is:

A if CONDITION else B

This means:

> If the condition is True, use A. Otherwise, use B.

Our expression is:

list(dict.fromkeys(x)) if isinstance(x, list) else x

We can visualize it as:

isinstance(x, list)
        ↓
   ┌────┴────┐
   ↓         ↓
 True      False
   ↓         ↓
list(       return
dict...     x
)

If x is a list:

x = ["C", "C", "C"]

isinstance(x, list) → True

So Python executes:

list(dict.fromkeys(x))

Result:

["C"]

If x is not a list:

x = NaN

isinstance(x, list) → False

So Python skips the left side and simply returns:

x

Result:

NaN

6. The Lambda Function

We can now understand the complete lambda:

lambda x: list(dict.fromkeys(x)) if isinstance(x, list) else x

The important thing to remember is that .apply() is what gives the lambda one entry at a time.

Conceptually:

Series
↓
.apply()
↓
Take one entry
↓
Give it to lambda as x
↓
Check if x is a list
↓
If list → remove duplicate deck letters
If not list → leave it unchanged

For example:

x = ["C", "C", "C"]

↓
isinstance(x, list)
↓
True
↓
list(dict.fromkeys(x))
↓
["C"]

Another example:

x = ["F", "G"]

↓
isinstance(x, list)
↓
True
↓
list(dict.fromkeys(x))
↓
["F", "G"]

Another example:

x = NaN

↓
isinstance(x, list)
↓
False
↓
return x
↓
NaN

7. Complete Expression

The complete operation is:

df_clean["Cabin"].str.findall(r"[A-G]").apply(
lambda x: list(dict.fromkeys(x)) if isinstance(x, list) else x
)

This performs the following steps:

df_clean["Cabin"]
↓
Select the Cabin column
↓
.str.findall(r"[A-G]")
↓
Extract all deck letters from each Cabin value
↓
.apply(...)
↓
Process each Cabin result one at a time
↓
lambda x
↓
Check whether x is a list
↓
If x is a list:
Remove duplicate deck letters
Convert the result back to a list
↓
If x is not a list:
Leave it unchanged

8. Examples

Before removing duplicates:

C23 C25 C27 → ["C", "C", "C"]

B57 B59 B63 B66 → ["B", "B", "B", "B"]

F G73 → ["F", "G"]

F E69 → ["F", "E"]

After removing duplicates:

C23 C25 C27 → ["C"]

B57 B59 B63 B66 → ["B"]

F G73 → ["F", "G"]

F E69 → ["F", "E"]

Missing Cabin:

NaN → NaN

Key Concepts
.apply()

.apply() processes the Series one entry at a time and passes each entry into the lambda as x.

lambda x

x represents the current individual entry being processed.

isinstance(x, list)

Checks whether the current entry is a list.

dict.fromkeys(x)

Uses the values as dictionary keys, which removes duplicate values because dictionary keys must be unique.

list(...)

Converts the result back into a list.

A if condition else B

Uses A when the condition is True and B when the condition is False.

One-Sentence Summary

.apply() gives the lambda one Cabin result at a time; isinstance(x, list) checks whether that result is a list, and if it is, list(dict.fromkeys(x)) removes duplicate deck letters and converts the result back into a list, while missing values such as NaN are left unchanged.

In [64]:
deck = df_clean["Cabin"].str.findall(r"[A-GT]").apply(
    lambda x: list(dict.fromkeys(x)) if isinstance(x, list) else x
)

In [65]:
df_clean["deck"] = deck

In [67]:
df_clean["deck"].head()

0    NaN
1    [C]
2    NaN
3    [C]
4    NaN
Name: deck, dtype: object

In [71]:
df_clean["deck"].value_counts(dropna=False)

#Normally, value_counts() does not count NaN values.

#so  dropna=False "Don't drop the NaN values. Count them too."

deck
NaN       687
[C]        59
[B]        47
[D]        33
[E]        32
[A]        15
[F]         9
[G]         4
[F, G]      3
[F, E]      1
[T]         1
Name: count, dtype: int64

We know Deck contains:

[C]
[B]
[F, G]
NaN
[]

And we found one [].

We want to find:

Which original Cabin value produced that empty list?

Start from the inside
df_clean["Deck"]

This gives us the Deck Series:

[C]
[B]
NaN
[F, G]
[]
...

Then:

.apply(lambda x: ...)

means:

Go through each Deck entry one at a time and give it to the lambda as x.

So x could be:

[C]

or:

[F, G]

or:

NaN

or:

[]
Now the lambda
lambda x: isinstance(x, list) and len(x) == 0

There are two conditions connected by and.

Condition 1
isinstance(x, list)

Asks:

Is x a list?

For:

[C]      → True
[F, G]   → True
[]       → True
NaN      → False
Condition 2
len(x) == 0

len() tells us how many items are inside something.

So:

[C]      → len = 1
[F, G]   → len = 2
[]       → len = 0

Therefore:

[] → len(x) == 0 → True
The and

We want both conditions to be true:

isinstance(x, list) and len(x) == 0

So:

x = [C]
Is it a list? YES
Does it have 0 items? NO
→ False
x = [F, G]
Is it a list? YES
Does it have 0 items? NO
→ False
x = []
Is it a list? YES
Does it have 0 items? YES
→ True
x = NaN
Is it a list? NO
→ False

So the lambda produces a Boolean filter:

False
False
False
True
False
...
Finally:
df_clean["Cabin"][condition]

This means:

Give me the Cabin values where our condition is True.

So we're going:

Deck
 ↓
find the one []
 ↓
get its row
 ↓
look back at Cabin
 ↓
show me the original Cabin value
In plain English
df_clean["Cabin"][df_clean["Deck"].apply(
    lambda x: isinstance(x, list) and len(x) == 0
)]

means:

"Look through the Deck column, find the entries that are empty lists, and then show me the corresponding original Cabin values."

In [70]:
df_clean["Deck"].value_counts(dropna=False)

Deck
NaN    890
T        1
Name: count, dtype: int64

Deck
NaN    890
T        1
Name: count, dtype: int64